test/parse/test_entity_declaration.rb

```diff
@@ -23,6 +23,239 @@ def parse(internal_subset)
+Malformed entity declaration
+Line: 1
+Position: 46
+Last 80 unconsumed characters:
+ valid-name \"% &\" > ]>
+            DETAIL
+          end
+
+          class TestExternalId < self
+            # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-ExternalID
+            class TestSystemLiteral < self
+              def test_no_quote_in_system
+                # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-SystemLiteral
+                exception = assert_raise(REXML::ParseException) do
+                  REXML::Document.new('<!DOCTYPE root [<!ENTITY valid-name SYSTEM invalid-system-literal > ]>')
+                end
+                assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 70
+Last 80 unconsumed characters:
+ valid-name SYSTEM invalid-system-literal > ]>
+                DETAIL
+              end
+
+              def test_no_quote_in_public
+                # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-SystemLiteral
+                exception = assert_raise(REXML::ParseException) do
+                  REXML::Document.new('<!DOCTYPE root [<!ENTITY valid-name PUBLIC "valid-pubid-literal" invalid-system-literal > ]>')
+                end
+                assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 92
+Last 80 unconsumed characters:
+ valid-name PUBLIC \"valid-pubid-literal\" invalid-system-literal > ]>
+                DETAIL
+              end
+            end
+
+            class TestPubidLiteral < self
+              def test_no_quote
+                # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-PubidLiteral
+                exception = assert_raise(REXML::ParseException) do
+                  REXML::Document.new('<!DOCTYPE root [<!ENTITY valid-name PUBLIC invalid-pubid-literal "valid-system-literal" > ]>')
+                end
+                assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 92
+Last 80 unconsumed characters:
+ valid-name PUBLIC invalid-pubid-literal \"valid-system-literal\" > ]>
+                DETAIL
+              end
+
+              def test_invalid_pubid_char
+                # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-PubidChar
+                exception = assert_raise(REXML::ParseException) do
+                  # U+3042 HIRAGANA LETTER A
+                  REXML::Document.new("<!DOCTYPE root [<!ENTITY valid-name PUBLIC \"\u3042\" \"valid-system-literal\" > ]>")
+                end
+                assert_equal(<<-DETAIL.force_encoding('utf-8').chomp, exception.to_s.force_encoding('utf-8'))
+Malformed entity declaration
+Line: 1
+Position: 76
+Last 80 unconsumed characters:
+ valid-name PUBLIC \"\u3042\" \"valid-system-literal\" > ]>
+                DETAIL
+              end
+            end
+          end
+
+          class TestNDataDeclaration < self
+            def test_no_quote
```

Comment:
Is "no quote" a suitable reason?
I think that `NDATA` is only the valid keyword.

Reply:
Ah, sorry. It is my mistake.